Various fixes and changes to the library code

PhoenixBound opened the pull request:
I've been trying to improve the reliability of [CoilSnake](https://github.com/pk-hack/CoilSnake), which consumes an old version of the library interface from exhal (replacing older bespoke code from the EB community). Part of that reliability work involved bubbling all errors up to Python, since I've seen at least one user have Python itself report an OOM error in CoilSnake's log window. I noticed that exhal's code assumed a few types of allocations never failed, which led to the creation of this PR.

The goal of this PR was to eliminate instances of undefined behavior, namely lack of handling for `malloc` and `calloc` failure, various reads and writes out of bounds, and other instances of undefined behavior caught by compiler instrumentation features (basically everything that can cause a crash with the compiler flags `-fsanitize=undefined -fsanitize=address -fsanitize-trap=all`). That hopefully gives users of the library more flexibility to handle and prevent errors to the extent that they need.

* The first commit and third commit upgrade dependencies. uthash gained support for not making out-of-memory errors fatal, and [memmem.c from musl-libc](https://git.musl-libc.org/cgit/musl/log/src/string/memmem.c) had minor bug fixes since April 2016, including a signed overflow issue that Undefined Behavior Sanitizer complains about. I don't have much to point out about these except that uthash's code isn't associated with a specific release, it uses the newest commit.
* The second commit and fifth commit have all the new error handling and bug fixes. The commit messages explain most of the details, but there are some things worth calling out in them:
    * The [documentation for uthash](https://troydhanson.github.io/uthash/userguide.html#_out_of_memory) suggested using C++ exceptions or C `setjmp`/`longjmp` as a potential method for handling allocation failures. The simplest way to accomplish that was a thread-local variable in the global scope. I'm not sure if I can use those with MSVC/Python/CoilSnake, though, since it's a C11 feature... I guess I still have to test that out. Worst case scenario, the thread-local variable can always be heap-allocated instead, or passed in by the caller somehow, or made into a normal global variable by #define-ing out the thread local keyword. I think the current state of things is the best option if it works, though.
    * When making the second commit, I added `(int)` casts to all of the COMBINE macros out of habit to cast it from `unsigned int` to `int`. It seems they're not actually necessary, though, since -Wall doesn't seem to warn about that conversion if I remove them.
    * The fifth commit includes a bug fix for `rle_check` unconditionally reading two bytes in order to check for 16-bit runs, which Address Sanitizer complained about. The `if` statement added inside that function seemed like the easiest way to handle it? The fix ended up having a different form from `backref_check`, though.
    * Some details of `setjmp` and what variables can and can't be preserved are very particular, and it seems like gcc's higher warning levels are blunt in comparison. I added a comment explaining why it seems okay to use `this` in the uthash error path of pack_context_alloc, based on the [rules for `setjmp` and `longjmp`](https://cppreference.com/w/c/program/setjmp.html), but gcc warns anyway that it might not be safe to use with a generic error message that isn't specific to setjmp.
* The fourth commit adds a bunch of `const`s to the public API (in compress.h), and to the internals of the library to avoid pointer casts (in compress.c). I'm not sure if this has any real benefits besides library users being able to leave out a few pointer casts, to remove their own `const`s before calling exhal code.

I made a [test program](https://github.com/PhoenixBound/exhal/commit/a0417a68c6beb717dcccb1f5827a15f7f28be864) to test all of these compression code changes on over 400 examples of compressed data taken from EarthBound. It takes a while to run a full test run, and Address Sanitizer slows things down a bit further, but I've completed a full test run on the current commit and confirmed that, at least on that small well-behaved set of files,

* there are no memory leaks caused by the malloc failure handling,
* there are no reads or writes out of bounds, at least when the packed and unpacked arrays are of size 65536,
* and no files in EarthBound round-trip from decompressed to compressed and back and have their contents or size change

Like I said, I haven't confirmed that I can use a thread-local variable in a Windows-Python interop context, but apart from that I *think* everything is okay to review? Putting this PR together took longer than expected...

PhoenixBound committed:
Upgrade uthash to latest commit (41c357f, after 2.3.0)

A feature to make allocation failures non-fatal was added in 2017. Some
other bugs and warnings have been fixed in it since then as well.

PhoenixBound committed:
Handle all malloc failures, don't shift into the sign bit of `int`

Previously, some failures of malloc/calloc were handled and others were
ignored. uthash would call exit when its allocations failed, which isn't
appropriate for library code. This adds setjmp-based error handling for
uthash, as recommended in its docs, and informs the caller that the
operation failed if allocating memory failed.

Replacing the sign bit of a signed integer with a left shift causes
undefined behavior in C (similar to other cases of signed integer
overflow). Compiler features like Undefined Behavior Sanitizer will
catch these errors and either report them or crash the program.
Switching bit shifting calculations to use an unsigned type fixes it.

PhoenixBound committed:
Update memmem.c from latest musl-libc

Namely, commit 593caa456309714402ca4cb77c3770f4c24da9da. This commit
also happens to be the one that fixes UndefinedBehaviorSanitizer's
warnings about signed overflow from shifting a byte left by 24.

PhoenixBound committed:
Make compress.c public API and internals const-correct

It seemed like a good idea at the time, though it's a breaking change
for any library consumers.

PhoenixBound committed:
Fix remaining bugs that ASan complains about

The first bug was a latent bug that was already there: rle_check always
checks for 16-bit RLE, including reading the two bytes for the potential
run, even if there's not two bytes available in the file to read. The
cleanest way to prevent that seemed to guard that entire section behind
a size check.

The other bugs were memory leaks that I introduced when handling malloc
failures.
In pack_context_alloc, if a malloc in uthash failed, it would leak the
element that was about to be inserted, because I forgot to free that
element before calling longjmp. This was fixed by actually freeing that
element.
In exhal_pack2, if pack_optimal failed, ctx would be leaked because I
immediately returned 0 before freeing the pack context. This was fixed
by letting a later point in the function handle return 0 instead.

devinacker commented:
Would it make sense to just make `hash_error_jmpbuf` a member of `pack_context_t`? It'd already be allocated by the time `setjmp` is called, and that way you don't need to worry about C11 support, thread-local semantics, etc.

The rest looks good to me.

PhoenixBound commented:
It would tie the uthash error handling code pretty closely to `pack_context_alloc` (since `recover_from_hash_oom` would need access to the pack_context_t containing the jmp_buf), but it *would* work, since that function's the only thing that needs uthash error handling...

Thinking more about it, since uthash does *everything* through macros and not functions, I might be able to leave out the setjmp stuff entirely and just use `goto` with some uniquely-enough-named labels. (And actually, adding an extra pack_context_t argument of `this` to `recover_from_hash_oom` only works for the same reason, since it needs to be able to access local variables to do that). I'm going to try the goto approach first when I get time.

PhoenixBound committed:
Switch to goto for error handling

setjmp/longjmp-based handling is more resilient to changes in uthash if
it were to suddenly start using functions in new versions of its
implementation, but this is easier and doesn't involve C11 keywords and
features.

PhoenixBound committed:
Remove unnecessary header
